internal/pkg/vm/internal/vmcontext/vmcontex2.go

```diff
@@ -0,0 +1,543 @@
+
+	// 3. build context
+	ctx := newInvocationContext(vm, imsg, fromActor, &gasTank)
+
+	// 4. invoke message
+	// Note: any errors raised by the invocation will not be trapped
+	// 		 errors are not expected by the spec on implicit messages, this messages must work
+	return ctx.invoke()
+}
+
+// applyMessage applies the message to the current state.
+func (vm *VM) applyMessage(msg *types.UnsignedMessage, onChainMsgSize uint32, miner address.Address) (message.Receipt, MinerPenaltyFIL) {
+	// This method does not actually execute the message itself,
+	// but rather deals with the pre/post processing of a message.
+	// (see: `invocationContext.invoke()` for the dispatch and execution)
+
+	// initiate gas tracking
+	gasTank := gas.NewTracker(msg.GasLimit)
+
+	// pre-send
+	// 1. charge for message existence
+	// 2. load sender actor
+	// 3. check message seq number
+	// 4. check if _sender_ has enough funds
+	// 5. increment message seq number
+	// 6. withheld maximum gas from _sender_
+	// 7. checkpoint state
+
+	// 1. charge for bytes used in chain
+	ok := gasTank.TryCharge(gascost.OnChainMessage(onChainMsgSize))
+	if !ok {
+		// Invalid message; insufficient gas limit to pay for the on-chain message size.
+		return message.Failure(exitcode.OutOfGas), gasTank.GasConsumed().Cost(msg.GasPrice)
+	}
+
+	// Dragons: from address need to be normalized here
+
+	// 2. load actor from global state
+	fromActor, err := vm.state.GetActor(context.Background(), msg.From)
+	// Review: when are errors surfaced? this can lead to invalid state transitions
+	if fromActor == nil || err != nil {
+		// Execution error; sender does not exist at time of message execution.
+		return message.Failure(exitcode.ActorNotFound), gasTank.GasConsumed().Cost(msg.GasPrice)
+	}
+
+	// 3. make sure this is the right message order for fromActor
+	if msg.CallSeqNum != fromActor.CallSeqNum {
+		// Execution error; invalid seq number.
+		return message.Failure(exitcode.InvalidCallSeqNum), gasTank.GasConsumed().Cost(msg.GasPrice)
+	}
+
+	// 4. Check sender balance (gas + value being sent)
+	gasLimitCost := msg.GasLimit.Cost(msg.GasPrice)
+	totalCost := msg.Value.Add(gasLimitCost)
+	if fromActor.Balance.LessThan(totalCost) {
+		// Execution error; sender does not have sufficient funds to pay for the gas limit.
+		return message.Failure(exitcode.InsufficientFunds), gasTank.GasConsumed().Cost(msg.GasPrice)
+	}
+
+	// 5. Increment sender CallSeqNum
+	fromActor.IncrementSeqNum()
+
+	// 6. Deduct gas limit funds from sender first
+	// Note: this should always succeed, due to the sender balance check above
+	// Note: after this point, we nede to return this funds back before exiting
+	vm.transfer(msg.From, address.BurntFundsAddress, gasLimitCost)
+
+	// 7. checkpoint state
+	// Even if the message fails, the following accumulated changes will be applied:
+	// - CallSeqNumber increment
+	// - sender balance withheld
+	// - (if did not previously exist, the new auto-created target actor)
```

Comment:
This last point is no longer true in the spec

Reply:
I'll have to to another pass on a separate PR to make sure everything matches the spec